include/nbl/video/IGPUCommandPool.h

```diff
@@ -4,38 +4,835 @@
+        static void operator delete  (void* ptr, std::size_t sz) = delete;
+        static void operator delete[](void* ptr, std::size_t sz) = delete;
+        static void operator delete  (void* ptr, std::size_t sz, std::align_val_t al) = delete;
+        static void operator delete[](void* ptr, std::size_t sz, std::align_val_t al) = delete;
+
+        inline uint32_t getSize() const { return m_size; }
+
+    protected:
+        ICommand(uint32_t size) : m_size(size) {}
+
+    private:
+        friend CCommandSegment;
+
+        const uint32_t m_size;
+    };
+
+    template <class CRTP>
+    class NBL_FORCE_EBO IFixedSizeCommand : public IGPUCommandPool::ICommand
+    {
     public:
-        enum E_CREATE_FLAGS : uint32_t
+        template <typename... Args>
+        static uint32_t calc_size(const Args&...)
+        {
+            return sizeof(CRTP);
+        }
+
+    protected:
+        IFixedSizeCommand() : ICommand(calc_size()) {}
+    };
+
+    class alignas(COMMAND_SEGMENT_ALIGNMENT) CCommandSegment
+    {
+    private:
+        struct header_t
         {
-            ECF_NONE = 0x00,
-            ECF_TRANSIENT_BIT = 0x01,
-            ECF_RESET_COMMAND_BUFFER_BIT = 0x02,
-            ECF_PROTECTED_BIT = 0x04
+            core::LinearAddressAllocator<uint32_t> m_commandAllocator;
+            CCommandSegment* m_next;
+        } header;
+
+    public:
+        static inline constexpr uint32_t STORAGE_SIZE = COMMAND_SEGMENT_SIZE - core::roundUp(sizeof(header_t), alignof(ICommand));
+
+        struct Iterator
+        {
+            ICommand* m_cmd = nullptr;
+            // uint32_t m_size = 0;
+            CCommandSegment* m_segment = nullptr;
+
+            inline bool operator!=(const Iterator& other) const { return m_cmd != other.m_cmd; }
+            // TODO(achal): Return Iterator&?
+            // inline void operator++()
+            // {
+            //     m_cmd = reinterpret_cast<ICommand*>(reinterpret_cast<uint8_t*>(m_cmd) + m_size);
+            //     m_size = ;
+            // }
         };
 
-        IGPUCommandPool(core::smart_refctd_ptr<const ILogicalDevice>&& dev, core::bitflag<E_CREATE_FLAGS> _flags, uint32_t _familyIx) : IBackendObject(std::move(dev)), m_flags(_flags), m_familyIx(_familyIx) {}
+        CCommandSegment()
+        {
+            header.m_commandAllocator = core::LinearAddressAllocator<uint32_t>(nullptr, 0u, 0u, COMMAND_SEGMENT_ALIGNMENT, STORAGE_SIZE);
+            header.m_next = nullptr;
 
-        core::bitflag<E_CREATE_FLAGS> getCreationFlags() const { return m_flags; }
-        uint32_t getQueueFamilyIndex() const { return m_familyIx; }
+            wipeNextCommandSize();
+        }
 
-        // OpenGL: nullptr, because commandpool doesn't exist in GL (we might expose the internal allocator in the future)
-        // Vulkan: const VkCommandPool*
-        virtual const void* getNativeHandle() const = 0;
+        template <typename Cmd, typename... Args>
+        void* allocate(const Args&... args)
```

Comment:
why not return `Cmd*` right away for convenience?

Reply:
It has been long, seems like we're going round in circles now

https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/59d39c0ea65dab6acbff4c24c4c1e21543b177e2#r863838228

https://github.com/Devsh-Graphics-Programming/Nabla/pull/345#discussion_r863819861

https://github.com/Devsh-Graphics-Programming/Nabla/pull/345/files/59d39c0ea65dab6acbff4c24c4c1e21543b177e2#r863838553

Reply:
> Let's clear something up, I'm not asking you to return a pointer to a constructed Cmd, I'm just asking you to not return a pointer to unitialized memory of type void, just a pointer to uninitialized memory of type Cmd.

Oh I didn't understand that. Its resolved now.